File: agents/video-auto/gen_grid.py
from PIL import Image
import os

slides_dir = "/workspace/agents/video-auto/slides"
output_dir = "/workspace/agents/video-auto/video"
os.makedirs(output_dir, exist_ok=True)

images = sorted([f for f in os.listdir(slides_dir) if f.endswith('.png')])
print(f"Found {len(images)} slides: {images}")

w, h = 1920, 1080
cols = 3
rows = (len(images) + cols - 1) // cols

canvas = Image.new('RGB', (w * cols, h * rows), (10, 10, 26))

for i, img_name in enumerate(images):
    img_path = f"{slides_dir}/{img_name}"
    try:
        img = Image.open(img_path).convert('RGB')
        img_ratio = img.width / img.height
        cell_ratio = w / h
        if img_ratio > cell_ratio:
            new_w, new_h = w, int(w / img_ratio)
        else:
            new_w, new_h = int(h * img_ratio), h
        img = img.resize((new_w, new_h), Image.LANCZOS)
        x = (w - new_w) // 2
        y = (h - new_h) // 2
        row, col = i // cols, i % cols
        canvas.paste(img, (col * w + x, row * h + y))
        print(f"  Added {i+1}: {img_name}")
    except Exception as e:
        print(f"  Error {img_name}: {e}")

grid_path = f"{slides_dir}/slides_grid.png"
canvas.save(grid_path, "PNG")
print(f"Grid saved: {grid_path} ({canvas.width}x{canvas.height})")

video_grid_path = f"{output_dir}/slides_grid.png"
canvas.save(video_grid_path, "PNG")
print(f"Copied to: {video_grid_path}")

File: agents/video-auto/gen_grid.js
const sharp = require('sharp');
const path = require('path');
const fs = require('fs');

const slidesDir = '/workspace/agents/video-auto/slides';
const outputDir = '/workspace/agents/video-auto/video';
const gridPath = '/workspace/agents/video-auto/slides/slides_grid.png';
const videoGridPath = '/workspace/agents/video-auto/video/slides_grid.png';

fs.mkdirSync(outputDir, { recursive: true });

const images = fs.readdirSync(slidesDir)
  .filter(f => f.endsWith('.png'))
  .sort();
console.log('Found', images.length, 'slides:', images);

const W = 1920, H = 1080;
const cols = 3;
const rows = Math.ceil(images.length / cols);

const cellW = W, cellH = H;

async function main() {
  const composites = [];

  for (let i = 0; i < images.length; i++) {
    const imgPath = path.join(slidesDir, images[i]);
    const row = Math.floor(i / cols);
    const col = i % cols;
    const left = col * cellW;
    const top = row * cellH;
    
    composites.push({
      input: imgPath,
      left: left,
      top: top
    });
    console.log(`  Slide ${i+1}: ${images[i]} at (${left}, ${top})`);
  }

  // Create the base dark canvas
  const canvasWidth = W * cols;
  const canvasHeight = H * rows;

  // First, create a dark background
  const bg = await sharp({
    create: {
      width: canvasWidth,
      height: canvasHeight,
      channels: 3,
      background: { r: 10, g: 10, b: 26 }
    }
  }).png().toBuffer();

  // Composite all images onto the dark background
  const result = await sharp(bg)
    .composite(composites)
    .png()
    .toBuffer();

  // Save to slides dir
  await sharp(result).toFile(gridPath);
  console.log('Grid saved to:', gridPath);

  // Also save to video dir
  await sharp(result).toFile(videoGridPath);
  console.log('Also saved to:', videoGridPath);

  // Report dimensions
  const meta = await sharp(result).metadata();
  console.log('Grid dimensions:', meta.width, 'x', meta.height);
}

main().catch(e => {
  console.error('Error:', e.message);
  process.exit(1);
});
